File: dotnet/src/CarespaceSDK/Exceptions/CarespaceException.cs
using System.Net;

namespace CarespaceSDK.Exceptions;

public class CarespaceException : Exception
{
    public HttpStatusCode? StatusCode { get; }
    public string? ErrorCode { get; }
    public object? ErrorDetails { get; }

    public CarespaceException(string message) : base(message)
    {
    }

    public CarespaceException(string message, Exception innerException) 
        : base(message, innerException)
    {
    }

    public CarespaceException(
        string message, 
        HttpStatusCode statusCode, 
        string? errorCode = null, 
        object? errorDetails = null) 
        : base(message)
    {
        StatusCode = statusCode;
        ErrorCode = errorCode;
        ErrorDetails = errorDetails;
    }

    public CarespaceException(
        string message, 
        HttpStatusCode statusCode, 
        Exception innerException, 
        string? errorCode = null, 
        object? errorDetails = null) 
        : base(message, innerException)
    {
        StatusCode = statusCode;
        ErrorCode = errorCode;
        ErrorDetails = errorDetails;
    }
}

public class CarespaceAuthenticationException : CarespaceException
{
    public CarespaceAuthenticationException(string message) 
        : base(message, HttpStatusCode.Unauthorized, "AUTHENTICATION_FAILED")
    {
    }

    public CarespaceAuthenticationException(string message, Exception innerException) 
        : base(message, HttpStatusCode.Unauthorized, innerException, "AUTHENTICATION_FAILED")
    {
    }
}

public class CarespaceAuthorizationException : CarespaceException
{
    public CarespaceAuthorizationException(string message) 
        : base(message, HttpStatusCode.Forbidden, "AUTHORIZATION_FAILED")
    {
    }

    public CarespaceAuthorizationException(string message, Exception innerException) 
        : base(message, HttpStatusCode.Forbidden, innerException, "AUTHORIZATION_FAILED")
    {
    }
}

public class CarespaceNotFoundException : CarespaceException
{
    public CarespaceNotFoundException(string message) 
        : base(message, HttpStatusCode.NotFound, "RESOURCE_NOT_FOUND")
    {
    }

    public CarespaceNotFoundException(string message, Exception innerException) 
        : base(message, HttpStatusCode.NotFound, innerException, "RESOURCE_NOT_FOUND")
    {
    }
}

public class CarespaceValidationException : CarespaceException
{
    public CarespaceValidationException(string message, object? validationErrors = null) 
        : base(message, HttpStatusCode.BadRequest, "VALIDATION_FAILED", validationErrors)
    {
    }

    public CarespaceValidationException(string message, Exception innerException, object? validationErrors = null) 
        : base(message, HttpStatusCode.BadRequest, innerException, "VALIDATION_FAILED", validationErrors)
    {
    }
}

public class CarespaceRateLimitException : CarespaceException
{
    public TimeSpan? RetryAfter { get; }

    public CarespaceRateLimitException(string message, TimeSpan? retryAfter = null) 
        : base(message, HttpStatusCode.TooManyRequests, "RATE_LIMIT_EXCEEDED")
    {
        RetryAfter = retryAfter;
    }

    public CarespaceRateLimitException(string message, Exception innerException, TimeSpan? retryAfter = null) 
        : base(message, HttpStatusCode.TooManyRequests, innerException, "RATE_LIMIT_EXCEEDED")
    {
        RetryAfter = retryAfter;
    }
}

public class CarespaceServerException : CarespaceException
{
    public CarespaceServerException(string message) 
        : base(message, HttpStatusCode.InternalServerError, "SERVER_ERROR")
    {
    }

    public CarespaceServerException(string message, Exception innerException) 
        : base(message, HttpStatusCode.InternalServerError, innerException, "SERVER_ERROR")
    {
    }
}

File: kotlin/carespace-sdk/src/commonMain/kotlin/com/carespace/sdk/exceptions/CarespaceException.kt
package com.carespace.sdk.exceptions

import io.ktor.http.*

/**
 * Base exception class for all Carespace SDK errors
 */
open class CarespaceException(
    message: String,
    val statusCode: HttpStatusCode? = null,
    val errorCode: String? = null,
    val errorDetails: Any? = null,
    cause: Throwable? = null
) : Exception(message, cause)

/**
 * Exception thrown when authentication fails
 */
class CarespaceAuthenticationException(
    message: String = "Authentication failed. Please check your API key.",
    cause: Throwable? = null
) : CarespaceException(
    message = message,
    statusCode = HttpStatusCode.Unauthorized,
    errorCode = "AUTHENTICATION_FAILED",
    cause = cause
)

/**
 * Exception thrown when authorization fails (insufficient permissions)
 */
class CarespaceAuthorizationException(
    message: String = "Access denied. You don't have permission to access this resource.",
    cause: Throwable? = null
) : CarespaceException(
    message = message,
    statusCode = HttpStatusCode.Forbidden,
    errorCode = "AUTHORIZATION_FAILED",
    cause = cause
)

/**
 * Exception thrown when a requested resource is not found
 */
class CarespaceNotFoundException(
    message: String = "The requested resource was not found.",
    cause: Throwable? = null
) : CarespaceException(
    message = message,
    statusCode = HttpStatusCode.NotFound,
    errorCode = "RESOURCE_NOT_FOUND",
    cause = cause
)

/**
 * Exception thrown when request validation fails
 */
class CarespaceValidationException(
    message: String = "Bad request. Please check your request data.",
    validationErrors: Any? = null,
    cause: Throwable? = null
) : CarespaceException(
    message = message,
    statusCode = HttpStatusCode.BadRequest,
    errorCode = "VALIDATION_FAILED",
    errorDetails = validationErrors,
    cause = cause
)

/**
 * Exception thrown when rate limit is exceeded
 */
class CarespaceRateLimitException(
    message: String = "Rate limit exceeded. Please try again later.",
    val retryAfterSeconds: Long? = null,
    cause: Throwable? = null
) : CarespaceException(
    message = message,
    statusCode = HttpStatusCode.TooManyRequests,
    errorCode = "RATE_LIMIT_EXCEEDED",
    cause = cause
)

/**
 * Exception thrown when server encounters an internal error
 */
class CarespaceServerException(
    message: String = "Internal server error occurred.",
    cause: Throwable? = null
) : CarespaceException(
    message = message,
    statusCode = HttpStatusCode.InternalServerError,
    errorCode = "SERVER_ERROR",
    cause = cause
)

/**
 * Exception thrown when network connectivity issues occur
 */
class CarespaceNetworkException(
    message: String = "Network error occurred. Please check your internet connection.",
    cause: Throwable? = null
) : CarespaceException(
    message = message,
    errorCode = "NETWORK_ERROR",
    cause = cause
)

/**
 * Exception thrown when request times out
 */
class CarespaceTimeoutException(
    message: String = "Request timed out. Please try again.",
    cause: Throwable? = null
) : CarespaceException(
    message = message,
    errorCode = "TIMEOUT_ERROR",
    cause = cause
)